pkg/rc/replication_controller.go

```diff
@@ -0,0 +1,243 @@
+	go func() {
+		for err := range errInChannel {
+			errOutChannel <- err
+		}
+		channelsClosed <- struct{}{}
+	}()
+
+	// If both channels I'm watching are closed,
+	// which would happen when my caller requests a quit
+	// (the quit gets forwarded to the rc watcher)
+	// close my output channel.
+	go func() {
+		<-channelsClosed
+		<-channelsClosed
+		close(channelsClosed)
+		close(errOutChannel)
+	}()
+
+	return errOutChannel
+}
+
+func (rc *replicationController) meetDesires() error {
+	rc.logger.NoFields().Infof("Meeting with desired replicas %d, disabled %v", rc.ReplicasDesired, rc.Disabled)
+
+	// If we're disabled, we do nothing, nor is it an error
+	// (it's a normal possibility to be disabled)
+	if rc.Disabled {
+		return nil
+	}
+
+	eligible, err := rc.eligibleNodes()
+	if err != nil {
+		return err
+	}
+	rc.logger.NoFields().Infof("Eligible to be on nodes %s", eligible)
+
+	current, err := rc.CurrentNodes()
+	if err != nil {
+		return err
+	}
+	rc.logger.NoFields().Infof("Currently on nodes %s", current)
+
+	if rc.ReplicasDesired > len(current) {
+		// TODO: With Docker or runc we would not be constrained to running only once per node.
+		// So it may be the case that we need to make the Scheduler interface smarter and use it here.
+		possible := difference(eligible, current)
+		toSchedule := rc.ReplicasDesired - len(current)
+
+		rc.logger.NoFields().Infof("Need to schedule %d nodes out of %s", toSchedule, possible)
+
+		if toSchedule > len(possible) {
+			return util.Errorf("Not enough nodes to meet desire: %d replicas desired, %d current, %d eligible", rc.ReplicasDesired, len(current), len(eligible))
+		}
+
+		for i := 0; i < toSchedule; i++ {
+			err := rc.schedule(possible[i])
+			if err != nil {
+				return err
+			}
+		}
+	} else if len(current) > rc.ReplicasDesired {
+		toUnschedule := len(current) - rc.ReplicasDesired
+		rc.logger.NoFields().Infof("Need to unschedule %d nodes out of %s", toUnschedule, current)
+
+		for i := 0; i < toUnschedule; i++ {
+			err := rc.unschedule(current[i])
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		rc.logger.NoFields().Debugln("Taking no action")
+	}
+
+	return nil
+}
+
+func (rc *replicationController) eligibleNodes() ([]string, error) {
+	return rc.scheduler.EligibleNodes(rc.Manifest, rc.NodeSelector)
+}
+
+func (rc *replicationController) CurrentNodes() ([]string, error) {
+	selector := rc.PodLabels.AsSelector()
+
+	pods, err := rc.podApplicator.GetMatches(selector, labels.POD)
+	if err != nil {
+		return []string{}, nil
+	}
+
+	// ID will be something like <nodename>/<podid>.
+	// We just want the nodename.
+	result := make([]string, len(pods))
+	for i, node := range pods {
+		parts := strings.SplitN(node.ID, "/", 2)
+		// If it so happens that node.ID contains no "/",
+		// then parts[0] is the entire string, so we're OK
+		result[i] = parts[0]
+	}
+	return result, nil
+}
+
+func (rc *replicationController) forEachLabel(node string, f func(id, k, v string) error) error {
```

Comment:
See, this is where I wish consul could do transactions spanning multiple KV operations. Comment is good idea, will have